src/main/java/build/buildfarm/worker/ReportResultStage.java

```diff
@@ -76,35 +76,61 @@ public void offer(OperationContext operationContext) {
     queue.offer(operationContext);
   }
 
-  private void updateActionResultStdOutputs(
+  private int updateActionResultStdOutputs(
       ActionResult.Builder resultBuilder,
-      ImmutableList.Builder<ByteString> contents) {
+      ImmutableList.Builder<ByteString> contents,
+      int inlineContentBytes) {
     ByteString stdoutRaw = resultBuilder.getStdoutRaw();
     if (stdoutRaw.size() > 0) {
-      CASInsertionControl control = worker.config.getStdoutCasControl();
-      boolean withinLimit = stdoutRaw.size() <= control.getLimit();
-      if (!withinLimit) {
+      CASInsertionPolicy policy = worker.config.getStdoutCasPolicy();
+      boolean withinLimit = inlineContentBytes + stdoutRaw.size() <= worker.config.getInlineContentLimit();
+      if (withinLimit) {
+        inlineContentBytes += stdoutRaw.size();
+      } else {
         resultBuilder.setStdoutRaw(ByteString.EMPTY);
       }
-      if (control.getPolicy() == CASInsertionControl.Policy.ALWAYS_INSERT ||
-          (!withinLimit && control.getPolicy() == CASInsertionControl.Policy.INSERT_ABOVE_LIMIT)) {
+      if (policy.equals(CASInsertionPolicy.ALWAYS_INSERT) ||
```

Comment:
Wait, so if we have NEVER_INSERT, and it's above limit, we just pretend there was no stdout at all? Ditto with the output files? That seems wrong. I think this should be an error instead.

Reply:
I'm fine with this being an error, but since NEVER_INSERT has this problem everywhere - it makes unusable configurations out of everything over the inline limit, which can be as low as 0 bytes - I would also be on board with removing NEVER_INSERT entirely as a policy.

Also, glad you pointed this out, because I missed a `contents.add` for `stdoutRaw`.

Reply:
Right, removing it entirely sounds good, too!
I missed that too -- I think it would be much less easy to miss if the code snippet was organized a bit differently. It took me a while to process the "contents are already in std*raw field to begin with, so here he just needs to CLEAR that field when the condition does NOT hold, instead of setting it when the condition does hold". It's the reverse of the canonical case, and is a bit hard to read.

How about restructuring it something like:

private int32 inlineOrDigest(
        ByteString content,
        CASInsertionPolicy policy,
        ImmutableList.Builder<ByteString> contents,
        int inlineContentBytes, int inlineContentLimit,
        Function<ByteString,Void> setRaw,
        Function<Digest,Void> setDigest) {
    boolean withinLimit = inlineContentBytes + content.size() <= inlineContentLimit;
    if (withinLimit) {
      setRaw(content);
      inlineContentBytes += content.size();
    }
    if (policy.equals(CASInsertionPolicy.ALWAYS_INSERT) ||
        (!withinLimit && policy.equals(CASInsertionPolicy.INSERT_ABOVE_LIMIT))) {
      contents.add(content);
      setDigest(Digests.computeDigest(content));
    }
    return inlineContentBytes;
}

and call this function for all three cases -- stdout, stderr, and output files?

Reply:
I was thinking something similar originally, making this static and adding a test or two for the logic preservation, but didn't want to completely jump that functional hurdle. I'll still need a special case to clear stdout and stderr, but yup, will apply.